sys/cddl/dev/dtrace/dtrace_ioctl.c

```diff
@@ -171,41 +193,48 @@ dtrace_ioctl(struct cdev *dev, u_long cmd, caddr_t addr,
 		 * the temporary buffer to be able to drop dtrace_lock()
 		 * across the copyout(), below.
 		 */
-		size = sizeof (dtrace_aggdesc_t) +
-		    (aggdesc.dtagd_nrecs * sizeof (dtrace_recdesc_t));
+		size = SIZEOF(dtrace_aggdesc) +
+		    (aggdesc.dtagd_nrecs * SIZEOF(dtrace_recdesc));
 
 		buf = kmem_alloc(size, KM_SLEEP);
 		dest = (uintptr_t)buf;
 
-		bcopy(&aggdesc, (void *)dest, sizeof (aggdesc));
-		dest += offsetof(dtrace_aggdesc_t, dtagd_rec[0]);
+		bcopy_aggdesc(&aggdesc, (void *)dest);
+		dest += OFFSETOF(dtrace_aggdesc, dtagd_rec[0]);
+
 
 		for (act = agg->dtag_first; ; act = act->dta_next) {
-			dtrace_recdesc_t rec = act->dta_rec;
+
+			/* using the following line doesn't work: it seems that
+			 * the compiler is doing some optimisation, and rec ends
+			 * up being all zero.
+			 * dtrace_recdesc_t rec = act->dta_rec;
+			 */
```

Comment:
This is extremely fishy. My instincts would be that the input code is wrong, not the compiler, and that this is papering over a bug in the DTrace code. Moreover, if the compiler _is_ at fault, this is a fragile workaround that could easily break in future. Either way, this should really be diagnosed and fixed properly, and this workaround should not be committed.

Reply:
I agree. I've already spent many hours on this without getting much...
The only thing I can see is that the compiler optimizes all the accesses to `rec` as zero.

Reply:
A good start would be to get a preprocessed version of the file that exhibits the bug so it's easy to poke at with different flags, look at the IR, cut out bits of code etc.

Reply:
Considering I know almost nothing about compilers, I might spend a huge amount of time on this without any result. Since I'm very short of time I might give precedence to other things.. maybe we can commit this with `#error` and the description? 
If you want to help, I'm happy to take a look at it together

Reply:
That's why I'd like the preprocessed source. I could revert that change and compile it myself, but I have no good way of knowing whether I've reproduced the bug, short of staring at the assembly. Whereas if you build CheriBSD and reproduce the bug with whatever you did to see it in the past, give the corresponding pre-processed source, compiler flags and version, Alex or I can then take it and forget about the DTrace side of things.